fuzzers/main.py

```diff
@@ -0,0 +1,842 @@
+        self.log_file.write(f'Current time: {datetime.datetime.now()}\n')
+        self.log_file.flush()
+
+        # Build the fuzzer runner
+        # TODO: actually build and have build time metrics for the fuzzer runner, but this must not
+        # block the pause/resume behavior
+        #build_start = time.monotonic()
+        #subprocess.check_call(
+        #    ['cargo', 'fuzz', 'build', self.target['runner']],
+        #    cwd=self.repo_dir / self.target['crate'],
+        #    stdout=self.log_file,
+        #    stderr=subprocess.STDOUT,
+        #)
+        #self.fuzz_build_time_metric.inc(time.monotonic() - build_start)
+
+    def start(self, corpus: Corpus) -> None:
+        """Start the fuzzer runner on corpus `Corpus`"""
+        print(f'Starting fuzzer for branch {self.branch} and '
+              f'target {self.target}, log is at {self.log_relpath}',
+              file=sys.stderr)
+
+        # Prepare the fuzz time metric
+        self.last_time = time.monotonic()
+
+        # Spin up the fuzzer process itself
+        # libfuzzer will kill the process if it takes more than -timeout number of seconds.
+        # nayduck can sigstop the fuzzing process for ~2 hours at most, so 8000s should be ok.
+        self.proc = subprocess.Popen(  # pylint: disable=consider-using-with
+            [
+                'cargo',
+                'fuzz',
+                'run',
+                self.target['runner'],
+                '--',
+                str(corpus.corpus_for(self.target)),
+                str(corpus.artifacts_for(self.target)),
+                f'-artifact_prefix={corpus.artifacts_for(self.target)}/',
+                '-timeout=8000',
+            ] + self.target['flags'],
+            cwd=self.repo_dir / self.target['crate'],
+            start_new_session=True,
+            stdout=self.log_file,
+            stderr=subprocess.STDOUT,
+        )
+
+    def poll(self) -> bool:
+        """Checks if the current process is still running. Returns True if it stopped"""
+        if self.proc.poll() is None:
+            new_time = time.monotonic()
+            self.fuzz_time_metric.inc(new_time - self.last_time)
+            self.last_time = new_time
+            return False
+        # else: Fuzz crash found
+        print(f'Fuzzer running {self.target} has stopped', file=sys.stderr)
+        self.fuzz_crashes_metric.inc()
+        return True
+
+    def report_crash(self, corpus: Corpus, bucket: gcs.Bucket) -> None:
+        # pylint: disable=too-many-locals
+
+        with open(self.log_fullpath, 'r', encoding='utf-8') as log_file_r:
+            log_lines = log_file_r.readlines()
+
+        artifact_pattern = f'Test unit written to {corpus.artifacts_for(self.target)}/'
+        artifact = '<failed detecting relevant artifact in log file>'
+        for line in log_lines[::-1]:
+            if artifact_pattern in line:
+                artifact = line.split(artifact_pattern)[1].strip()
+                break
```

Comment:
If you wanna be a hax0r than this can also be written as:

```python
artifact = next((line.split(artifact_pattern)[1].strip()
                 for line in log_lines[::-1]
                 if artifact_pattern in line),
                '<failed detecting relevant artifact in log file>')
```